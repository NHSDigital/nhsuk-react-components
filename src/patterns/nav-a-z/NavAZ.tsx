import React, { HTMLProps } from 'react';
import classNames from 'classnames';
import type { AsElementLink } from '../../util/types/LinkTypes';

const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';

const processLetters = (
  children: React.ReactNode,
  fullAlphabet: boolean | undefined,
  removedLetters: Array<string> | undefined,
  disabledLetters: Array<string> | undefined,
  letters: Array<string> | undefined,
) => {
  const letterMap = (letter: string) => {
    if (removedLetters && removedLetters.includes(letter)) {
      return null;
    }
    if (disabledLetters && disabledLetters.includes(letter)) {
      return <DisabledItem key={letter}>{letter}</DisabledItem>;
    }
    return (
      <LinkItem key={letter} href={`#${letter}`}>
        {letter}
      </LinkItem>
    );
  };
  if (children) {
    return children;
  }
  if (fullAlphabet) {
    return alphabet.split('').map(letterMap);
  }
  if (letters) {
    return letters.map(letterMap);
  }
  return null;
};

interface NavAZProps extends HTMLProps<HTMLDivElement> {
  fullAlphabet?: boolean;
  removedLetters?: Array<string>;
  disabledLetters?: Array<string>;
  letters?: Array<string>;
}

interface NavAZ extends React.FC<NavAZProps> {
  LinkItem: React.FC<AsElementLink<HTMLAnchorElement>>;
  DisabledItem: React.FC<HTMLProps<HTMLSpanElement>>;
}

const NavAZ: NavAZ = ({
  className,
  children,
  fullAlphabet,
  removedLetters,
  disabledLetters,
  letters,
  ...rest
}) => (
  <nav className={classNames('nhsuk-nav-a-z', className)} {...rest}>
    <ol className="nhsuk-nav-a-z__list">
      {processLetters(children, fullAlphabet, removedLetters, disabledLetters, letters)}
    </ol>
  </nav>
);

const LinkItem: React.FC<AsElementLink<HTMLAnchorElement>> = ({
  className,
  asElement: Component = 'a',
  ...rest
}) => (
  <li className="nhsuk-nav-a-z__item">
    <Component className={classNames('nhsuk-nav-a-z__link', className)} {...rest} />
  </li>
);

const DisabledItem: React.FC<HTMLProps<HTMLSpanElement>> = ({ className, ...rest }) => (
  <li className="nhsuk-nav-a-z__item">
    <span className={classNames('nhsuk-nav-a-z__link--disabled', className)} {...rest} />
  </li>
);

NavAZ.LinkItem = LinkItem;
NavAZ.DisabledItem = DisabledItem;

export default NavAZ;
