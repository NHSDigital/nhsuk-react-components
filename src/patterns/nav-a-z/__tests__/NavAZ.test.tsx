/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable react/jsx-key */
import React from 'react';
import { shallow, mount } from 'enzyme';
import NavAZ from '..';

const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';

describe('NavAZ', () => {
  it('matches snapshot', () => {
    const element = shallow(<NavAZ />);
    expect(element).toMatchSnapshot('NavAZ');
    element.unmount();
  });

  it('renders fullAlphabet', () => {
    const element = mount(<NavAZ fullAlphabet />);
    const expectedElements = alphabet.split('').map((letter) => (
      <a key={letter} className="nhsuk-nav-a-z__link">
        {letter}
      </a>
    ));
    expect(element.containsAllMatchingElements(expectedElements)).toBeTruthy();
    element.unmount();
  });

  it('renders fullAlphabet with letters removed', () => {
    const element = mount(<NavAZ fullAlphabet removedLetters={['A', 'B']} />);
    expect(
      element.containsAnyMatchingElements([
        <a className="nhsuk-nav-a-z__link">A</a>,
        <a className="nhsuk-nav-a-z__link">B</a>,
      ]),
    ).toBeFalsy();
    element.unmount();
  });

  it('renders fullAlphabet with letters disabled', () => {
    const element = mount(<NavAZ fullAlphabet disabledLetters={['A', 'B']} />);
    expect(
      element.containsAnyMatchingElements([
        <a className="nhsuk-nav-a-z__link">A</a>,
        <a className="nhsuk-nav-a-z__link">B</a>,
      ]),
    ).toBeFalsy();
    expect(
      element.containsAllMatchingElements([
        <span className="nhsuk-nav-a-z__link--disabled">A</span>,
        <span className="nhsuk-nav-a-z__link--disabled">B</span>,
      ]),
    ).toBeTruthy();
    element.unmount();
  });

  it('passes through children', () => {
    const element = mount(
      <NavAZ>
        <p id="child">Child</p>
      </NavAZ>,
    );
    expect(element.containsMatchingElement(<p id="child">Child</p>)).toBeTruthy();
    element.unmount();
  });

  it('renders individual letters', () => {
    const element = mount(<NavAZ letters={['A', 'B']} />);
    expect(
      element.containsAllMatchingElements([
        <a className="nhsuk-nav-a-z__link">A</a>,
        <a className="nhsuk-nav-a-z__link">B</a>,
      ]),
    ).toBeTruthy();
    element.unmount();
  });
});
