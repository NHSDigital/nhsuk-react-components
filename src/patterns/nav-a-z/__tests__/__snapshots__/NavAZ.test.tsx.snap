// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavAZ matches snapshot: NavAZ 1`] = `
<nav
  className="nhsuk-nav-a-z"
>
  <ol
    className="nhsuk-nav-a-z__list"
  />
</nav>
`;
